src/dali/plugin/input/rest/binance_com.py

```diff
@@ -0,0 +1,774 @@
+            #   'addressFrom': None,
+            #   'tag': None,
+            #   'tagTo': None,
+            #   'tagFrom': None,
+            #   'type': 'deposit',
+            #   'amount': 0.00999800,
+            #   'currency': 'PAXG',
+            #   'status': 'ok',
+            #   'updated': None,
+            #   'internal': False,
+            #   'fee': None
+            # }
+
+            for deposit in crypto_deposits:
+                self._process_transfer(deposit, intra_transactions)
+
+            # If user made more than 1000 transactions in a 90 day period we need to shrink the window.
+            if len(crypto_deposits) < 1000:
+                current_start = current_end + 1
+                current_end = current_start + 7776000000
+            else:
+                # Binance sends latest record first ([0])
+                # CCXT sorts by timestamp, so latest record is last ([999])
+                current_start = int(crypto_deposits[999][_TIMESTAMP]) + 1  # times are inclusive
+                current_end = current_start + 7776000000
+
+        # Process actual fiat deposits (no limit on the date range)
+        # Fiat deposits can also be pulled via CCXT fetch_deposits by cycling through legal_money
+        # Using the underlying api call is faster for Binance.
+        fiat_deposits = self.client.sapiGetFiatOrders(params=({_TRANSACTIONTYPE: 0, _STARTTIME: self.start_time_ms, _ENDTIME: now_time}))
+        #    {
+        #      "code": "000000",
+        #      "message": "success",
+        #      "data": [
+        #        {
+        #          "orderNo": "25ced37075c1470ba8939d0df2316e23",
+        #          "fiatCurrency": "EUR",
+        #          "indicatedAmount": "15.00",
+        #          "amount": "15.00",
+        #          "totalFee": "0.00",
+        #          "method": "card",
+        #          "status": "Failed",
+        #          "createTime": 1627501026000,
+        #          "updateTime": 1627501027000
+        #        }
+        #      ],
+        #      "total": 1,
+        #      "success": True
+        #    }
+        if _DATA in fiat_deposits:
+            for deposit in fiat_deposits[_DATA]:
+                if deposit[_STATUS] == "Completed":
+                    self._process_deposit(deposit, in_transactions)
+
+    def _process_gains(self, in_transactions: List[InTransaction]) -> None:
+
+        ### Regular Dividends from Staking (including Eth staking) and Savings (Lending)
+
+        # We need milliseconds for Binance
+        current_start = self.start_time_ms
+        now_time = int(datetime.datetime.now().timestamp()) * 1000
+
+        # We will pull in 30 day periods. This allows for 16 assets with daily dividends.
+        current_end = current_start + 2592000000
+
+        while current_start < now_time:
+
+            # CCXT doesn't have a standard way to pull income, we must use the underlying API call

```

Comment:
Just for my information, since I haven't used CCXT: how do you distinguish a standard CCXT call from an underying API one? Is it the `sapi` prefix?

Reply:
CCXT makes all the underlying api calls available by simply using their endpoint as a function. So, `sapiGetAssetAssetDividend` can be used for the `GET /sapi/v1/asset/assetDividend` endpoint. The `sapi` prefix actually refers to the 'spot api', which also covers margin trades, mining and savings. There are also apis for different future markets - `dapi`, `vapi`, `fapi`.

Reply:
Oh, I see... I was under the impression that CCXT also had an API that is exchange-independent (meaning that it works everywhere), which may not cover all the functionality of every exchange but will cover most bases. Is that assumption incorrect?

Reply:
It does. All of the `fetch` (public API) and `fetch_my_` (private API) functions are standard CCXT functions that can be used for all exchanges. It's just that Binance.com has a lot more features than the average exchange. For instance, I don't know of many other exchanges that have mining pools connected directly to the exchange.

I'm surprised they don't have a standard function for dividends/staking etc... though. It's missing, so we will have to use the API endpoints for each exchange. 

For all US-based exchanges, we can use the `fetch_deposits` for fiat deposits. I had to use `FiatOrders` and `FiatPayments` endpoints since Binance.com handles a lot of fiat pairs not just one (like all US-based exchanges, and even JPN-based exchanges)

You can also 'ask' CCXT if an exchange has a standard function:
```
if exchange.has['fetchMyTrades']:
    exchange.fetch_my_trades(symbol=None, since=None, limit=None, params={})
```

Reply:
Got it, thanks for the details: makes sense. It's a good approach and it sounds like we'll have to use `exchange.has` a lot in the superclass.